sdk/resourcemanager/azure-resourcemanager-resources/src/main/java/com/azure/resourcemanager/resources/fluentcore/arm/ResourceUtils.java

```diff
@@ -156,23 +157,30 @@ public static String nameFromResourceId(String id) {
      * @return the default api version to use
      */
     public static String defaultApiVersion(String id, Provider provider) {
-        String resourceType = resourceTypeFromResourceId(id).toLowerCase(Locale.ROOT);
-        // Exact match
+        if (id == null || provider == null) {
+            return null;
+        }
+        ResourceId resourceId = ResourceId.fromString(id);
+        String resourceTypeWithoutNamespace = getFullResourceTypeWithoutNamespace(resourceId);
+        // exact match first
         for (ProviderResourceType prt : provider.resourceTypes()) {
-            if (prt.resourceType().equalsIgnoreCase(resourceType)) {
+            // There is an edge case that two resource types share the same child resource type name, so parent type name check is needed.
+            // e.g. "dnsForwardingRulesets/virtualNetworkLinks" and "privateDnsZones/virtualNetworkLinks"
+            if (prt.resourceType().equalsIgnoreCase(resourceTypeWithoutNamespace)) {
                 return prt.defaultApiVersion() == null ? prt.apiVersions().get(0) : prt.defaultApiVersion();
             }
         }
-        // child resource, e.g. sites/config
+
+        // relaxed match, in case of edge cases(e.g. namespace included in resource type)
+        String fullResourceType = resourceId.fullResourceType();
         for (ProviderResourceType prt : provider.resourceTypes()) {
-            if (prt.resourceType().toLowerCase(Locale.ROOT).contains("/" + resourceType)) {
+            if (fullResourceType.contains(prt.resourceType())) {
```

Comment:
This match logic (type from resource contains type from prt) seems the opposite or previous logic (type from prt contains type from resource, maybe not the full type)?

Reply:
Yeah, it's the opposite. 

Yet `prt.resourceType()` may be incomplete? Wondering what the behaviour is. I'm still trying to figure out the relax match situation, is it like:
full resource type -> `level1/level2/level3/level4`
prt.resourceType() -> `level1/level4` ?

Reply:
You can take a look on the list of prt. `level1/level4` seems not make sense. But I am not sure if `level4` could happen.

You can also check with .NET. I assume they have similar logic (they may also have a cache :-))

BTW, you need consistent `.toLowerCase(Locale.ROOT)` here.

Reply:
Though I am still a bit unsure about match `fullResourceType.contains(prt.resourceType())`.

It seems already covers the "parent case" fallback, (I assume if parent id has exact match, here the relax match would already matches it?)
https://github.com/Azure/azure-sdk-for-java/blob/9c354f639a9b260e8461c8f17e6b8602a5b43e70/sdk/resourcemanager/azure-resourcemanager-resources/src/main/java/com/azure/resourcemanager/resources/fluentcore/arm/ResourceUtils.java#L183

Reply:
Any better idea? May be "exact match" first, "parent match" second, then the last chance "relax match"?

Reply:
Offline discussion: remove the `relax match` first, deal with it when needed.